clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -609,24 +732,32 @@ fn get_body_search_span(cx: &LateContext<'_>) -> Option<Span> {
     let body = cx.enclosing_body?;
     let map = cx.tcx.hir();
     let mut span = map.body(body).value.span;
-    let mut maybe_global_var = false;
-    for (_, node) in map.parent_iter(body.hir_id) {
-        match node {
+    let mut is_const_or_static = false;
+    for (_, parent_node) in map.parent_iter(body.hir_id) {
+        match parent_node {
             Node::Expr(e) => span = e.span,
             Node::Block(_) | Node::Arm(_) | Node::Stmt(_) | Node::Local(_) => (),
             Node::Item(hir::Item {
                 kind: hir::ItemKind::Const(..) | ItemKind::Static(..),
                 ..
-            }) => maybe_global_var = true,
-            Node::Item(hir::Item {
-                kind: hir::ItemKind::Mod(_),
-                span: item_span,
+            })
+            | Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Const(..),
                 ..
-            }) => {
-                span = *item_span;
+            })
+            | Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Const(..),
+                ..
+            }) => is_const_or_static = true,
+            // Parent is a `mod xxx { .. }``
+            Node::Item(item) => {
+                match item.kind {
+                    ItemKind::Mod(_) | ItemKind::Trait(..) | ItemKind::Impl(..) => span = item.span,
+                    _ => (),
+                }
                 break;
             },
-            Node::Crate(mod_) if maybe_global_var => {
+            Node::Crate(mod_) if is_const_or_static => {
                 span = mod_.spans.inner_span;
             },
             _ => break,
```

Comment:
Not caused by your PR but I found this function a bit confusing, what do you think of

```rust
fn get_body_search_span(cx: &LateContext<'_>) -> Option<Span> {
    let body = cx.enclosing_body?;
    for (_, parent_node) in cx.tcx.hir().parent_iter(body.hir_id) {
        match parent_node {
            Node::Item(hir::Item {
                kind: hir::ItemKind::Const(..) | ItemKind::Static(..),
                ..
            })
            | Node::ImplItem(hir::ImplItem {
                kind: hir::ImplItemKind::Const(..),
                ..
            })
            | Node::TraitItem(hir::TraitItem {
                kind: hir::TraitItemKind::Const(..),
                ..
            }) => {},
            Node::Item(item) => return Some(item.span),
            Node::Crate(mod_) => return Some(mod_.spans.inner_span),
            _ => {},
        }
    }
    None
}
```

Reply:
wow this is interesting, I would never realize that they  does the same thing, really appreciate your effort in this~ :smile:

Reply:
I think it's awesome, I got confused first but I only just rename a couple variables then add my code on top of it and call it a day, so,  thank you :smile: